pkg/solana/chainwriter/chain_writer.go

```diff
@@ -0,0 +1,324 @@
+  - Example: A pre-defined fee payer or system account.
+
+2. **AccountLookup**:
+  - Dynamically derived from input args using a specified location path (e.g., `user.walletAddress`).
+  - If the lookup table is pre-loaded, the address is fetched from `derivedTableMap`.
+
+3. **PDALookups**:
+  - Generates Program Derived Addresses (PDA) by combining a derived public key with one or more seeds.
+  - Seeds can be `AddressSeeds` (public keys from the input args) or `ValueSeeds` (byte arrays).
+  - Ensures there is only one public key if multiple seeds are provided.
+
+### Error Handling:
+- Errors are wrapped with the `debugID` for easier tracing.
+*/
+// GetAddresses resolves account addresses from various `Lookup` configurations to build the required `solana.AccountMeta` list
+// for Solana transactions.
+func GetAddresses(ctx context.Context, args any, accounts []Lookup, derivedTableMap map[string]map[string][]*solana.AccountMeta, reader client.Reader) ([]*solana.AccountMeta, error) {
+	var addresses []*solana.AccountMeta
+	for _, accountConfig := range accounts {
+		meta, err := accountConfig.Resolve(ctx, args, derivedTableMap, reader)
+		if err != nil {
+			return nil, err
+		}
+		addresses = append(addresses, meta...)
+	}
+	return addresses, nil
+}
+
+func (s *SolanaChainWriterService) FilterLookupTableAddresses(
+	accounts []*solana.AccountMeta,
+	derivedTableMap map[string]map[string][]*solana.AccountMeta,
+	staticTableMap map[solana.PublicKey]solana.PublicKeySlice,
+) map[solana.PublicKey]solana.PublicKeySlice {
+	filteredLookupTables := make(map[solana.PublicKey]solana.PublicKeySlice)
+
+	// Build a hash set of account public keys for fast lookup
+	usedAccounts := make(map[string]struct{})
+	for _, account := range accounts {
+		usedAccounts[account.PublicKey.String()] = struct{}{}
+	}
+
+	// Filter derived lookup tables
+	for _, innerMap := range derivedTableMap {
+		for innerIdentifier, metas := range innerMap {
+			tableKey, err := solana.PublicKeyFromBase58(innerIdentifier)
+			if err != nil {
+				continue
+			}
+
+			// Collect public keys that are actually used
+			var usedAddresses solana.PublicKeySlice
+			for _, meta := range metas {
+				if _, exists := usedAccounts[meta.PublicKey.String()]; exists {
+					usedAddresses = append(usedAddresses, meta.PublicKey)
+				}
+			}
+
+			// Add to the filtered map if there are any used addresses
+			if len(usedAddresses) > 0 {
+				filteredLookupTables[tableKey] = usedAddresses
+			}
+		}
+	}
+
+	// Filter static lookup tables
+	for tableKey, addresses := range staticTableMap {
+		var usedAddresses solana.PublicKeySlice
+		for _, staticAddress := range addresses {
+			if _, exists := usedAccounts[staticAddress.String()]; exists {
+				usedAddresses = append(usedAddresses, staticAddress)
+			}
+		}
+
+		// Add to the filtered map if there are any used addresses
+		if len(usedAddresses) > 0 {
+			filteredLookupTables[tableKey] = usedAddresses
+		}
+	}
+
+	return filteredLookupTables
+}
+
```

Comment:
Thinking in terms of users, we should have a comment here so they don't have to dive into the implementation to understand it. Something like: 

```suggestion
// SubmitTransaction builds, encodes, and enqueues a transaction using the provided program
// configuration and method details. It relies on the configured IDL, account lookups, and
// lookup tables to gather the necessary accounts and data. The function retrieves the latest
// blockhash and assigns it to the transaction, so callers do not need to provide one.
//
// Submissions and retries are handled by the underlying transaction manager. If a “debug ID”
// location is configured, SubmitTransaction extracts it from the provided `args` and attaches
// it to errors for easier troubleshooting. Only the first debug ID it encounters will be used.
//
// Parameters:
//   - ctx: The context for cancellation and timeouts.
//   - contractName: Identifies which Solana program config to use from `s.config.Programs`.
//   - method: Specifies which method config to invoke within the chosen program config.
//   - args: Arbitrary arguments that are encoded into the transaction payload and/or used for dynamic address lookups.
//   - transactionID: A unique identifier for the transaction, used for tracking within the transaction manager.
//   - toAddress: The on-chain address (program ID) to which the transaction is directed.
//   - meta: Currently unused; included for interface compatibility.
//   - value: Currently unused; included for interface compatibility.
//
// Returns:
//   An error if any stage of the transaction preparation or enqueueing fails. A nil return
//   indicates that the transaction was successfully submitted to the transaction manager.
```

Reply:
Awesome! adding this in